test/Microsoft.AspNetCore.SignalR.Tests/HubEndpointTests.cs

```diff
@@ -1333,6 +1332,89 @@ public static IEnumerable<object[]> StreamingMethodAndHubProtocols
             }
         }
 
+        [Fact]
+        public async Task DoesNotWritePingMessagesIfSufficientOtherMessagesAreSent()
+        {
+            var serviceProvider = CreateServiceProvider(services =>
+                services.Configure<HubOptions>(options =>
+                    options.KeepAliveInterval = TimeSpan.FromMilliseconds(100)));
+            var endPoint = serviceProvider.GetService<HubEndPoint<MethodHub>>();
+
+            using (var client = new TestClient(false, new JsonHubProtocol()))
+            {
+                var endPointLifetime = endPoint.OnConnectedAsync(client.Connection).OrTimeout();
+
+                await client.Connected.OrTimeout();
+
+                // Echo a bunch of stuff, waiting 10ms between each, until 500ms have elapsed
+                DateTime start = DateTime.UtcNow;
+                while ((DateTime.UtcNow - start).TotalMilliseconds <= 500.0)
+                {
+                    await client.SendInvocationAsync("Echo", "foo").OrTimeout();
+                    await Task.Delay(10);
+                }
+
+                // Shut down
+                client.Dispose();
+
+                await endPointLifetime.OrTimeout();
+
+                // We shouldn't have any ping messages
+                HubMessage message;
+                var counter = 0;
+                while ((message = await client.ReadAsync()) != null)
+                {
+                    counter += 1;
+                    Assert.IsNotType<PingMessage>(message);
+                }
+                Assert.InRange(counter, 1, 50);
+            }
+        }
+
+        [Fact]
+        public async Task WritesPingMessageIfNothingWrittenWhenKeepAliveIntervalElapses()
+        {
+            var serviceProvider = CreateServiceProvider(services =>
+                services.Configure<HubOptions>(options =>
+                    options.KeepAliveInterval = TimeSpan.FromMilliseconds(100)));
+            var endPoint = serviceProvider.GetService<HubEndPoint<MethodHub>>();
+
+            using (var client = new TestClient(false, new JsonHubProtocol()))
+            {
+                var endPointLifetime = endPoint.OnConnectedAsync(client.Connection).OrTimeout();
+                await client.Connected.OrTimeout();
+
+                // Wait 500 ms, but make sure to yield some time up to unblock concurrent threads
+                // This is useful on AppVeyor because it's slow enough to end up with no time
+                // being available for the endpoint to run.
+                for (var i = 0; i < 5; i += 1)
+                {
+                    // Tick the heartbeat feature manually, 10 times, every 10ms
+                    for(var j = 0; j < 10; j += 1)
+                    {
+                        client.Connection.TickHeartbeat();
+                        await Task.Yield();
```

Comment:
What's this for?

Reply:
Artifact from when the heartbeat was in a background timer. In my experience I'd seen cases where the timer fired but the async code it called didn't because xunit's sync context was blocking us. I'll remove it now though, should be unnecessary.